RAG_supporters/clustering/test_keyword_clustering.py

```diff
@@ -0,0 +1,400 @@
+"""
+Unit tests for keyword_clustering module.
+
+Tests cover:
+- KeywordClusterer initialization
+- Clustering operations (fit, predict)
+- Topic descriptor extraction
+- Assignment operations (hard/soft modes)
+- Configuration persistence (save/load)
+- State management
+"""
+
+import json
+import tempfile
+from pathlib import Path
+
+import numpy as np
+import pytest
+
+from .keyword_clustering import (KeywordClusterer,
+                                 cluster_keywords_from_embeddings)
+
+
+@pytest.fixture
+def sample_embeddings():
+    """Create sample keyword embeddings for testing."""
+    np.random.seed(42)
+    n_keywords = 20
+    embedding_dim = 10
+
+    # Create embeddings with 3 distinct clusters
+    embeddings = {}
+    for i in range(n_keywords):
+        if i < 7:
+            base = np.array([1.0, 0.0, 0.0] + [0.0] * (embedding_dim - 3))
+        elif i < 14:
+            base = np.array([0.0, 1.0, 0.0] + [0.0] * (embedding_dim - 3))
+        else:
+            base = np.array([0.0, 0.0, 1.0] + [0.0] * (embedding_dim - 3))
+
+        embedding = base + np.random.normal(0, 0.05, embedding_dim)
+        embedding = embedding / np.linalg.norm(embedding)
+        embeddings[f"keyword_{i}"] = embedding
+
+    return embeddings
+
+
+@pytest.fixture
+def fitted_clusterer(sample_embeddings):
+    """Create a fitted clusterer for testing."""
+    clusterer = KeywordClusterer(n_clusters=3, random_state=42)
+    clusterer.fit(sample_embeddings)
+    return clusterer
+
+
+class TestKeywordClustererInit:
+    """Test KeywordClusterer initialization."""
+
+    def test_init_parameters(self):
+        """Test initialization with default and custom parameters."""
+        # Test default parameters
+        clusterer = KeywordClusterer()
+        assert clusterer.algorithm == "kmeans"
+        assert clusterer.n_clusters == 8
+        assert clusterer.random_state == 42
+        assert clusterer.keywords is None
+        assert clusterer.embeddings_matrix is None
+        assert clusterer.cluster_labels is None
+
+        # Test default assignment config
+        assert clusterer._default_assignment_mode == "soft"
+        assert clusterer._default_threshold == 0.1
+        assert clusterer._default_metric == "cosine"
+
+        # Test custom parameters
+        clusterer = KeywordClusterer(
+            algorithm="bisecting_kmeans", n_clusters=5, random_state=123
+        )
+        assert clusterer.algorithm == "bisecting_kmeans"
```

Comment:
assert statements should left message if failed:
example:
assert len(results) == 5, f"Expected 5 results, got {len(results)}"